steward/src/cellars/cellar.rs

```diff
@@ -0,0 +1,228 @@
+    types::H160,
+};
+use steward_abi::cellar::{
+    AddPositionCall, CellarCalls, PushPositionCall, RebalanceCall, RemovePositionCall,
+    SetDepositLimitCall, SetHoldingPositionCall, SetLiquidityLimitCall, SetRebalanceDeviationCall,
+    SetShareLockPeriodCall, SetStrategistPayoutAddressCall, SetWithdrawTypeCall, SwapPositionsCall,
+};
+use steward_proto::steward::cellar::{swap_params::Params::*, Function, SwapParams};
+
+use crate::{
+    error::{Error, ErrorKind},
+    utils::{self, sp_call_error, sp_call_parse_address, string_to_u256},
+};
+
+use super::log_cellar_call;
+
+const CELLAR_NAME: &str = "Cellar";
+
+pub fn get_encoded_call(function: Function, cellar_id: String) -> Result<Vec<u8>, Error> {
+    match function {
+        Function::AddPosition(params) => {
+            log_cellar_call(CELLAR_NAME, &AddPositionCall::function_name(), &cellar_id);
+            let call = AddPositionCall {
+                index: params.index.into(),
+                position: sp_call_parse_address(params.position)?,
+            };
+
+            Ok(CellarCalls::AddPosition(call).encode())
+        }
+        Function::PushPosition(params) => {
+            log_cellar_call(CELLAR_NAME, &PushPositionCall::function_name(), &cellar_id);
+            let call = PushPositionCall {
+                position: sp_call_parse_address(params.position)?,
+            };
+
+            Ok(CellarCalls::PushPosition(call).encode())
+        }
+        Function::RemovePosition(params) => {
+            log_cellar_call(
+                CELLAR_NAME,
+                &RemovePositionCall::function_name(),
+                &cellar_id,
+            );
+            let call = RemovePositionCall {
+                index: params.index.into(),
+            };
+
+            Ok(CellarCalls::RemovePosition(call).encode())
+        }
+        Function::SetHoldingPosition(params) => {
+            log_cellar_call(
+                CELLAR_NAME,
+                &SetHoldingPositionCall::function_name(),
+                &cellar_id,
+            );
+            let call = SetHoldingPositionCall {
+                new_holding_position: sp_call_parse_address(params.new_holding_position)?,
+            };
+
+            Ok(CellarCalls::SetHoldingPosition(call).encode())
+        }
+        Function::Rebalance(params) => {
+            log_cellar_call(CELLAR_NAME, &RebalanceCall::function_name(), &cellar_id);
+            let swap_params =
+                encode_swap_params(params.params.ok_or_else(|| {
+                    ErrorKind::SPCallError.context("swap params cannot be empty")
+                })?)?;
+
+            info!("encoded: {:?}", hex::encode(&swap_params));
+
+            let call = RebalanceCall {
+                from_position: sp_call_parse_address(params.from_position)?,
+                to_position: sp_call_parse_address(params.to_position)?,
+                assets_from: string_to_u256(params.assets_from)?,
+                exchange: (params.exchange - 1) as u8,
```

Comment:
Curious about the `-1` here...wouldn't that make the enum values incorrect? It looks like here:

https://github.com/PeggyJV/cellar-contracts/blob/03ef9c1fc37ed5630964ad05ca3a62b4cbcba358/src/modules/swap-router/SwapRouter.sol#L24-L26

That the enum doesn't have an "unspecified" for the zero value, if those two are assigned to zero and one respectively. Why did we add an extra type and do this subtraction?

Reply:
Protobuf forces you to have a zero value for unspecified so this is the necessary conversion

Reply:
We may want to add some comments to this effect to make it clear why the code is doing this.